src/database-sqlite3.cpp

```diff
@@ -136,6 +137,23 @@ void Database_SQLite3::endSave() {
 	SQLRES(sqlite3_step(m_stmt_end), SQLITE_DONE,
 		"Failed to commit SQLite3 transaction");
 	sqlite3_reset(m_stmt_end);
+
+	int total_size;
+	int checkpointed;
+	SQLOK(sqlite3_wal_checkpoint_v2(m_database, NULL,
+			SQLITE_CHECKPOINT_PASSIVE, &total_size, &checkpointed),
+		"failed to checkpoint database");
+	if (total_size >= 0 && checkpointed >= 0) {
+		int new_wal_backlog = total_size - checkpointed;
+		// Normally, sqlite would checkpoint every 1000 pages.
+		if (m_wal_backlog / 10000 < new_wal_backlog / 10000)
+			(new_wal_backlog >= 100000 ? errorstream : warningstream)
+				<< "SQLITE3: WAL backlog exceeds "
+				<< (new_wal_backlog / 10000) * 10000
+				<< " pages - is another process using the database ?"
+				<< std::endl;
+		m_wal_backlog = new_wal_backlog;
+	}
```

Comment:
IIUC, any checkpoint, whether it happens at commit or upon explicit request, will checkpoint as many pages as possible. That would mean that if a commit is more than 1000 pages, the subsequent explicit checkpoint would be a no-op. 

So the code effectively already operates as if auto-checkpoint is disabled. I might as well disable it explicitly...

I'll also add a separate checkpointing thread. That should eliminate disk IO delays in the server and emerge threads altogether.